File: src/01.py
import aoc
from itertools import cycle

@aoc.main('01')
def main(indata: str) -> tuple[int,int]:
  nums, seen, s = [int(l) for l in indata.split('\n')], set(), 0
  for x in cycle(nums):
    s += x
    if s in seen:
      return sum(nums), s
    seen.add(s)

if __name__ == "__main__":
  main()

File: src/bin/01.rs
use hashbrown::HashMap;

#[aoc::main(01)]
fn main(input: &str) -> (i64, i64) {
    let (mut l1, mut l2): (Vec<_>, Vec<_>) = input.split('\n').map(|l| {
        let (a, b) = l.split_once(' ').unwrap();
        (a.trim().parse::<i64>().unwrap(), b.trim().parse::<i64>().unwrap())
    }).unzip();
    l1.sort();
    l2.sort();

    let mut count = HashMap::new();
    for &b in &l2 {
        *count.entry(b).or_insert(0) += 1;
    }

    let p1 = l1.iter().zip(&l2).map(|(a, b)| (a - b).abs()).sum();
    let p2 = l1.iter().map(|&a| a * count.get(&a).unwrap_or(&0)).sum();
    (p1, p2)
}
